### PR TITLE
Tests: Use one Circle CI container per build

### PR DESCRIPTION
**Pull request by islemaster**
Circle configuration changes necessary to switch from using two Circle CI containers per build down to one.

**Timing:** Based on current build timing information, I estimate this will add about 20 minutes to our average build time, but double the number of builds we can run in parallel, which is probably the more efficient way to use our containers.

### CODE REVIEW

File: circle.yml [hunk 1]
```diff
@@ -23,7 +23,7 @@ machine:
     - redis
 checkout:
   post:
-    - case $CIRCLE_NODE_INDEX in 1) ./bin/circle/merge-eyes-baselines ;; *) echo "Skipping step." ;; esac:
+    - ./bin/circle/merge-eyes-baselines:
```
Comment (islemaster) [inline circle.yml]: Leaving this here to keep the dream alive.  The baseline merge is actually disabled [within the `merge-eyes-baselines` script](https://github.com/code-dot-org/code-dot-org/blob/staging/bin/circle/merge-eyes-baselines#L6-L7).

File: circle.yml [hunk 2]
```diff
@@ -72,14 +72,14 @@ dependencies:
...
-        case $CIRCLE_NODE_INDEX in 0) echo "Skipping" ;; *) echo "
+        echo "
         use_dynamo_tables: false
         use_dynamo_properties: false
         no_https_store: true
         override_dashboard: \"localhost-studio.code.org\"
         override_pegasus: \"localhost.code.org\"
         dashboard_port: 3000
-        pegasus_port: 3000" >> locals.yml ;; esac
+        pegasus_port: 3000" >> locals.yml
```
Comment (islemaster) [inline circle.yml]: We could technically merge this step with the one above it, but leaving it separate for now so it's easy to undo in the future if we'd like.

File: lib/rake/circle.rake [hunk 3]
```diff
@@ -83,7 +83,7 @@ namespace :circle do
     end
     RakeUtils.system_stream_output 'until $(curl --output /dev/null --silent --head --fail http://localhost-studio.code.org:3000); do sleep 5; done'
     Dir.chdir('dashboard/test/ui') do
-      container_features = `find ./features -name '*.feature' | sort | awk "NR % (${CIRCLE_NODE_TOTAL} - 1) == (${CIRCLE_NODE_INDEX} - 1)"`.split("\n").map {|f| f[2..-1]}
+      container_features = `find ./features -name '*.feature' | sort`.split("\n").map {|f| f[2..-1]}
```
Comment (islemaster) [inline lib/rake/circle.rake:86]: Have to avoid divide-by-zero here.

File: circle.yml [hunk 4]
```diff
@@ -111,7 +111,15 @@ database:
...
-    - case $CIRCLE_NODE_INDEX in 0) bundle exec rake lint:javascript circle:run_tests ;; *) bundle exec rake circle:run_ui_tests ;; esac:
+    - bundle exec rake lint:javascript:
+        parallel: true
+        timeout:
+            7200
+    - bundle exec rake circle:run_tests:
+        parallel: true
+        timeout:
+            7200
+    - bundle exec rake circle:run_ui_tests:
```
Comment (davidsbailey) [inline circle.yml]: This means eyes, ui tests and lint will be run in parallel with each other, correct? This seems right, just confirming.

Comment (islemaster) [reply circle.yml]: No - `parallel: true` just means "this step is allowed to run on all containers, not just container zero." ([documentation](https://circleci.com/docs/1.0/parallel-manual-setup/#splitting-your-test-suite)) It's not very intuitive.  We'll be running these in sequence, in part because of memory concerns.

**Comment by islemaster** [comment]
Okay, first test failure here.  It looks like I did break some test environment setting.  Investigating...

TL;DR:
`Expected /https:\/\/test\-studio\.code\.org to match https://localhost-studio.code.org`

```
DeviseMailerTest
  test_invitation_instructions                                    FAIL (0.06s)
Minitest::Assertion:         Expected /https:\/\/test\-studio\.code\.org\/users\/invitation\/accept\?invitation_token=faketoken/ to match "<p>Welcome testuser166@example.com.xx!</p>\r\n\r\n<p>A Code Studio account has been created for you. Please click on the link below to confirm your account details and begin using it.</p>\r\n\r\n<p><a href=\"https://localhost-studio.code.org/users/invitation/accept?invitation_token=faketoken\">Confirm your account</a></p>\r\n\r\n<p>If you have any questions, do not hesitate to contact <a href=\"mailto:pd@code.org\">pd@code.org</a>.</p>\r\n\r\n<p>\r\nThanks,<br/>\r\nThe Code.org PD Team\r\n</p>\r\n".
        test/mailers/devise_mailer_test.rb:17:in `block in <class:DeviseMailerTest>'
        test/testing/setup_all_and_teardown_all.rb:22:in `run'
```

**Comment by islemaster** [comment]
Also seeing some of these:

```
  test_should_get_show_of_ECSPD_if_signed_in                     ERROR (0.07s)
ActiveRecord::RecordNotFound:         ActiveRecord::RecordNotFound: Couldn't find Script with id|name=ECSPD
```

Maybe unit tests do depend on some seeded content?

**Comment by islemaster** [comment]
Hmm... with changed db setup, still seeing some URL changes:
```
  test_code_org_root_path_in_prod                                 FAIL (0.00s)
Minitest::Assertion:         Expected: "//code.org"
          Actual: "//localhost.code.org:3000"
```
I wonder what else needs to be changed?

**Comment by islemaster** [comment]
Just realized - it's [this bit](https://github.com/code-dot-org/code-dot-org/pull/16118/files#diff-29944324a3cbf9f4bd0162dfe3975d88L79) of circle.yml that used to only apply to container 1.  I need this to only apply to UI tests.  I wonder how I can do that later in the build.

**Comment by islemaster** [comment]
:cry: I'm not sure we're going to be able to do this - it's like there's some mutually-exclusive environment setup between our UI tests and our dashboard unit tests.

**Comment by davidsbailey** [comment]
Thanks for investigating this Brad and I'm sorry it's turned out to be such a pain. It appears that the current problem is that dashboard db is not getting seeded with the levels needed by the UI tests. I'm going to try iterating on this locally to see if I can figure out the right steps. If you decide to pick this up again, let's sync up so that we're not duplicating effort.

**Comment by davidsbailey** [comment]
Good news Brad, this is working now after fixing the rake versioning problem, which was causing `rake seed:ui_test` to fail. Somehow running `bundle exec rake ...` from within `dashboard/` was picking up the wrong version of rake, but running it from the top level picks up the right version. 

### performance

We've gotten 3 consecutive green circle runs, running the full test suite in 2 containers each. The circle runs completed in:
* 1:53:15 https://circleci.com/gh/code-dot-org/code-dot-org/41569
* 2:27:27 https://circleci.com/gh/code-dot-org/code-dot-org/41566
* 1:54:06 https://circleci.com/gh/code-dot-org/code-dot-org/41523 (different branch, same bits)

Though these times may seem surprisingly long, this is only about 20-29 minutes slower than the last few 2-container staging builds, which completed in [1:38:16, 1:33:23, 1:54:41, 1:33:08, 1:35:54, 1:32:42, 1:30:13, 2:02:03].

### merge plan

I'm torn between merging this now to release container pressure, or trying to fix the huge slowdown we've seen over the past few days first. **I'm leaning toward merging this and going down to 1 container today** since we can still use the run times of the `run_tests` and `run_ui_tests` sections as benchmarks. We'll just have to remember  that this PR will artificially decrease the amount of time taken by `rake install` because it moves some of the work into `rake circle:seed_ui_test`.

**Comment by islemaster** [comment]
Awesome, thanks for picking this up Dave. I'm also in favor of merging
sooner rather than later. I think Brent is making progress on the slowdown
too.
